openshift-kube-apiserver/authorization/minimumkubeletversion/minimum_kubelet_version.go

```diff
@@ -0,0 +1,82 @@
+package minimumkubeletversion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	nodelib "github.com/openshift/library-go/pkg/config/node"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+)
+
+var minimumKubeletVersion *semver.Version
+
+func SetMinimumKubeletVersion(version string) {
+	if version == "" {
+		return
+	}
+	v := semver.MustParse(version)
+	minimumKubeletVersion = &v
+}
+
+type minimumKubeletVersionAuth struct {
+	nodeIdentifier nodeidentifier.NodeIdentifier
+	nodeInformer   corev1informers.NodeInformer
+}
+
+func NewMinimumKubeletVersion(nodeIdentifier nodeidentifier.NodeIdentifier,
+	nodeInformer corev1informers.NodeInformer,
+) authorizer.Authorizer {
+
+	return &minimumKubeletVersionAuth{
+		nodeIdentifier: nodeIdentifier,
+		nodeInformer:   nodeInformer,
+	}
+}
+
+func (m *minimumKubeletVersionAuth) Authorize(ctx context.Context, attrs authorizer.Attributes) (authorizer.Decision, string, error) {
```

Comment:
Is anything making sure that the node informer has synced before its cache is accessed?

Reply:
hm when it's [constructed](https://github.com/openshift/kubernetes/blob/master/pkg/controlplane/apiserver/config.go#L159) it has a resync period of 10m. I am not seeing anything that updates it more frequently. I do see it being used elsewhere https://github.com/openshift/kubernetes/blob/master/pkg/kubeapiserver/authorizer/config.go#L105 which was why I originally thought it safe to use. Do you think I should verify it's synced?

Reply:
It must be synced to use safely. If it's not, it could end up (for example) seeing an empty list and authorizing a request that should not be authorized.

Reply:
just checking: it seems like you're talking about checking HasSynced() to make sure it's initialized, rather than ensuring a resync has happened recently enough to ensure new node objects that have been added don't race against the creation of this object, right?

Reply:
actually, for this we're just calling a `Get` on the node name of the node we're attempting to authorize: https://github.com/openshift/kubernetes/pull/2104/files#diff-b7bdcb41a84b0a21cc83959481df9b457492b2e10b6d1350dff1278c8489bf44R71-R74

if we haven't synced yet, we'll get denied (because the node object wouldn't exist yet AFAICT). We could check if the informer has synced to clarify the error on why the node object doesn't exist yet but functionally we shouldn't approve a node joining without its node object is propagated in the informer so I'm not sure we need to

Reply:
> could check if the informer has synced to clarify the error on why the node object doesn't exist yet

OK, since it's necessary to look at the Node to come to an opinion on the request, I'd check if it's synced at the very top and short-circuit with a specific error. That's defensive against any future changes to how the lister is accessed, and it heads off confused bug reports with users saying "but my node _does_ exist!"

Reply:
(failing to ensure that informers have synced before reading their cache is a common source of bugs for us which is why I am hesitant to just be satisfied by failing closed)

Reply:
alrighty added!